service/src/main/java/bio/terra/pipelines/dependencies/leonardo/AppUtils.java

```diff
@@ -0,0 +1,199 @@
+  }
+
+  int appComparisonFunction(ListAppResponse a, ListAppResponse b, List<AppType> appTypeList) {
+    // First criteria: Prefer apps with the expected app type.
+    // NB: Negative because lower index is better
+    int appTypeScoreA = -appTypeList.indexOf(a.getAppType());
+    int appTypeScoreB = -appTypeList.indexOf(b.getAppType());
+    if (appTypeScoreA != appTypeScoreB) {
+      return appTypeScoreA - appTypeScoreB;
+    }
+    // If there is a WDS app type present, do this check; does not apply to cromwell app-types
+    if (a.getAppType() == AppType.WDS || b.getAppType() == AppType.WDS) {
+      // Second criteria: Prefer apps with the expected app type name
+      int nameScoreA =
+          Objects.equals(a.getAppName(), "wds-%s".formatted(imputationConfiguration.workspaceId()))
+              ? 1
+              : 0;
+      int nameScoreB =
+          Objects.equals(b.getAppName(), "wds-%s".formatted(imputationConfiguration.workspaceId()))
+              ? 1
+              : 0;
+      if (nameScoreA != nameScoreB) {
+        return nameScoreA - nameScoreB;
+      }
+    }
+
+    // Third criteria: tie-break on whichever is older
+    return OffsetDateTime.parse(a.getAuditInfo().getCreatedDate())
+        .compareTo(OffsetDateTime.parse(b.getAuditInfo().getCreatedDate()));
+  }
+
+  /**
+   * Invokes logic to determine the appropriate app for WDS and CROMWELL. If app is not running, a
+   * URL will not be present, in this case we return empty string Note: This logic is similar to how
+   * DataTable finds WDS app in Terra UI
+   *
+   * <p>(<a
+   * href="https://github.com/DataBiosphere/terra-ui/blob/ac13bdf3954788ca7c8fd27b8fd4cfc755f150ff/src/libs/ajax/data-table-providers/WdsDataTableProvider.ts#L94-L147">...</a>)
+   */
+  ListAppResponse findBestAppForAppType(List<ListAppResponse> apps, AppType appType)
+      throws DependencyNotAvailableException {
+    // WDS and Cromwell apps look for Kubernetes deployment statuses (such as RUNNING or
+    // PROVISIONING), expressed by
+    // Leo
+    // See here for specific enumerations --
+    // https://github.com/DataBiosphere/leonardo/blob/develop/core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/kubernetesModels.scala
+    // look explicitly for a RUNNING app named 'wds-${app.workspaceId}' or
+    // 'terra-app-<random_uuid>' -- if app is healthy and
+    // running, there should only be one app RUNNING
+    // an app may be in the 'PROVISIONING', 'STOPPED', 'STOPPING', which can still be deemed as an
+    // OK state for Leonardo apps
+    List<AppType> appTypeList;
+
+    // WDS and CROMWELL apps will get their proxy urls from a different group of app types (located
+    // in application.yml). Because of this, we need to identify the app type we need a url for, and
+    // get the url from the correct group of relevant apps.
+    if (appType.equals(AppType.WDS)) {
+      appTypeList = leonardoServerConfiguration.wdsAppTypeNames();
+    } else {
```

Comment:
we're currently assuming the only app types we'll ever encounter are WDS and CROMWELL... would it be better to explicitly add `if (appType.equals(AppType.CROMWELL))` and then the final else throws an "AppType %s not supported" error?

Reply:
I realize now that we'll likely adjust this method a bit as we go on, so fine to leave this as is for now